src/platform/x86/x86_64/kernel.c

```diff
@@ -0,0 +1,251 @@
+#include "assert.h"
+#include "kernel.h"
+#include "multiboot2.h"
+#include "string.h"
+#include "boot_comp.h"
+#include "mem_layout.h"
+#include "chal_cpu.h"
+
+#include <captbl.h>
+#include <retype_tbl.h>
+#include <component.h>
+#include <thd.h>
+#include <chal_plat.h>
+
+#define ADDR_STR_LEN 8
+
+boot_state_t initialization_state = INIT_BOOTED;
+
+void
+boot_state_transition(boot_state_t from, boot_state_t to)
+{
+	assert(initialization_state == from);
+	assert((to - from) == 1); /* transitions must be linear */
+	initialization_state = to;
+}
+
+struct mem_layout glb_memlayout;
+volatile int cores_ready[NUM_CPU];
+
+extern u8_t end; /* from the linker script */
+
+#define MEM_KB_ONLY(x) (((x) & ((1 << 20) - 1)) >> 10)
+#define MEM_MB_ONLY(x) ((x) >> 20)
+
+/*
+ * X86_64-FIXME:
+ * Change these two external variable into non-external
+ */
+extern u8_t _binary_constructor_start, _binary_constructor_end;
+
+void
+kern_memory_setup(u64_t mboot_addr, u64_t mboot_magic)
+{
+	unsigned int i = 0, wastage = 0;
+	struct multiboot_tag *tag;
+
+	glb_memlayout.allocs_avail = 1;
+
+	if (mboot_magic != MULTIBOOT2_BOOTLOADER_MAGIC) {
+		die("multiboot magic not correct\n");
+	}
+
+	if (mboot_addr & 7)
+    {
+	  die("mboot unligned mbi\n");
+    }
+
+	glb_memlayout.kern_end = &end;
+	assert((u64_t)&end % RETYPE_MEM_NPAGES * PAGE_SIZE == 0);
+
+	printk("Initial component found:\n");
+	/* These values have to be higher-half addresses */
+	glb_memlayout.mod_start = &_binary_constructor_start;
+	glb_memlayout.mod_end   = &_binary_constructor_end;
+
+	/* FIXME: set mod start and end temporarily in order to test */
+	//glb_memlayout.mod_start = &end + 1024;
```

Comment:
remove commented out code.

If this FIXME is relevant, it should likely be fixed, I'd imagine.